torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp

```diff
@@ -504,9 +670,14 @@ std::vector<Value*> FixupONNXControlflowNode(Node* n, int opset_version) {
 void FixupONNXControlflowNodeOutputs(Node* n) {
   switch (n->kind()) {
     case ::c10::onnx::Loop: {
+      // inputs (0, 1) are (i, cond).
       auto loop_carried_output_size = n->blocks().at(0)->inputs().size() - 2;
       for (auto i : c10::irange(n->outputs().size())) {
-        auto type = n->blocks().at(0)->outputs().at(i + 1)->type();
+        // Get the type from the input rather than the output
+        // to handle the case where a block input is Optional but the
+        // output is not (i.e. if the loop executes > 0 times, the
+        // output will not be None).
+        auto type = n->blocks().at(0)->inputs().at(i + 2)->type();
```

Comment:
Tests are failing at this line in the case where `ConvertSequenceDependencies` is triggered because that breaks the expectation that the block inputs contain all the block outputs. I'm looking into how to fix this. Let me know if you have ideas.

Reply:
Fixed!